In [62]:
import numpy as np  # Linear algebra library
import pandas as pd  # Data manipulation and analysis
import matplotlib.pyplot as plt  # Data visualization
import seaborn as sns  # Statistical plot visualization
import datetime

In [2]:
df1 = pd.read_csv("data.csv")  # Read-in data
df1.head()  # Preview data

Make       Model  Year             Engine Fuel Type  Engine HP  \
0  BMW  1 Series M  2011  premium unleaded (required)      335.0   
1  BMW    1 Series  2011  premium unleaded (required)      300.0   
2  BMW    1 Series  2011  premium unleaded (required)      300.0   
3  BMW    1 Series  2011  premium unleaded (required)      230.0   
4  BMW    1 Series  2011  premium unleaded (required)      230.0   

   Engine Cylinders Transmission Type     Driven_Wheels  Number of Doors  \
0               6.0            MANUAL  rear wheel drive              2.0   
1               6.0            MANUAL  rear wheel drive              2.0   
2               6.0            MANUAL  rear wheel drive              2.0   
3               6.0            MANUAL  rear wheel drive              2.0   
4               6.0            MANUAL  rear wheel drive              2.0   

                         Market Category Vehicle Size Vehicle Style  \
0  Factory Tuner,Luxury,High-Performance      Compact         Coupe   
1                     Luxury,Performance      Compact   Convertible   
2                Luxury,High-Performance      Compact         Coupe   
3                     Luxury,Performance      Compact         Coupe   
4                                 Luxury      Compact   Convertible   

   highway MPG  city mpg  Popularity   MSRP  
0           26        19        3916  46135  
1           28        19        3916  40650  
2           28        20        3916  36350  
3           28        18        3916  29450  
4           28        18        3916  34500

In [ ]:
# Get size of dataframe (rows, columns)
df1.shape

(11914, 16)

In [7]:
# Explore contents of certain columns that may be unnecessary
for col in df1.columns:
    print(f'Number of unique values in {col}: {df1[col].nunique()}')

Number of unique values in Make: 48
Number of unique values in Model: 915
Number of unique values in Year: 28
Number of unique values in Engine Fuel Type: 10
Number of unique values in Engine HP: 356
Number of unique values in Engine Cylinders: 9
Number of unique values in Transmission Type: 5
Number of unique values in Driven_Wheels: 4
Number of unique values in Number of Doors: 3
Number of unique values in Market Category: 71
Number of unique values in Vehicle Size: 3
Number of unique values in Vehicle Style: 16
Number of unique values in highway MPG: 59
Number of unique values in city mpg: 69
Number of unique values in Popularity: 48
Number of unique values in MSRP: 6049


In [46]:
# Identify columns for missing data
df1.isnull().sum()

Make                    0
Model                   0
Year                    0
Engine Fuel Type        3
Engine HP              69
Engine Cylinders       30
Transmission Type       0
Driven_Wheels           0
Number of Doors         6
Market Category      3742
Vehicle Size            0
Vehicle Style           0
highway MPG             0
city mpg                0
Popularity              0
MSRP                    0
dtype: int64

- Since our dataset contains 11,914 entries, we can approach NaNs by simply removing those rows
- Certain columns (such as Market Category) can be removed entirely since it likely has no bearing on MSRP, therefore we don't have to worry about those NaNs

In [49]:
# Select columns to be removed
keep_columns = ['Make', 'Model', 'Year', 'Engine HP', 'Transmission Type', 'Driven_Wheels', 'highway MPG', 'city mpg', 'MSRP']
df2 = df1[keep_columns]

In [54]:
df2.shape

(11914, 9)

In [52]:
# Recheck for Missing/NaN values
df2.isnull().sum()

Make                  0
Model                 0
Year                  0
Engine HP            69
Transmission Type     0
Driven_Wheels         0
highway MPG           0
city mpg              0
MSRP                  0
dtype: int64

In [57]:
# For Engine HP, we'll just drop the rows with missing values since it's less than 1% of our total dataset
df3 = df2.dropna()
df3.isnull().sum()

Make                 0
Model                0
Year                 0
Engine HP            0
Transmission Type    0
Driven_Wheels        0
highway MPG          0
city mpg             0
MSRP                 0
dtype: int64

In [58]:
# Check for duplicates
df3_duplicates = df3[df3.duplicated()]
print(df3_duplicates.shape)

(990, 9)


In [61]:
# Drop duplicate rows
df4 = df3.drop_duplicates()
df4.shape

(10855, 9)

In [71]:
# Create a new feature for the age of the car
df4.loc[:, 'Age'] =  datetime.date.today().year - df4['Year']